File: 236-lowest-common-ancestor-of-a-binary-tree/236-lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        i,ans=self.lca(root,p,q)
        return ans
    def lca(self,root,p,q):
        if(root):
            i,a=self.lca(root.left,p,q)
            if(a):
                return [0,a]
            j,a=self.lca(root.right,p,q)
            if(a):
                return [0,a]
            i+=j
            if(root==p):
                i+=1
            if(root==q):
                i+=1
            if(i==2):
                return [0,root]
            return [i,None]
        return [0,None]
        

File: 236-lowest-common-ancestor-of-a-binary-tree/236-lowest-common-ancestor-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        auto ans=lca(root,p,q);
        return ans.second;
    }
    pair<int,TreeNode*> lca(TreeNode *root,TreeNode *p,TreeNode* q){
        if(root){
            auto a=lca(root->left,p,q);
            if(a.second){
                return a;
            }
            auto b=lca(root->right,p,q);
            if(b.second){
                return b;
            }
            a.first+=b.first;
            if(p==root){
                a.first+=1;
            }
            if(q==root){
                a.first+=1;
            }
            if(a.first==2){
                return {0,root};
            }
            return {a.first,NULL};
        }
        return {0,NULL};
    }
};